x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                }
+                break;
+            case JwtRealmSettings.CLIENT_AUTHORIZATION_TYPE_NONE:
+            default:
+                if (Strings.hasText(actualSecret)) {
+                    throw new Exception("Rejected client authentication for type [" + type + "] due to present secret.");
+                }
+                break;
+        }
+        return true;
+    }
+
+    public static boolean validateSignedJwt(
+        final SignedJWT signedJwt,
+        final JWKSet jwkSet,
+        final String allowedIssuer,
+        final List<String> allowedAudiences,
+        final List<String> allowedSignatureAlgorithms,
+        final long allowedClockSkewSeconds
+    ) throws Exception {
+        final JWSHeader jwsHeader = signedJwt.getHeader();
+        final JWSAlgorithm jwtHeaderAlgorithm = jwsHeader.getAlgorithm();
+        final JOSEObjectType jwtHeaderType = jwsHeader.getType();
+        final String jwsHeaderKeyID = jwsHeader.getKeyID();
+        // Validate claims
+        final JWTClaimsSet jwtClaimsSet = signedJwt.getJWTClaimsSet();
+        final Date now = new Date();
+        final Date iat = jwtClaimsSet.getIssueTime();
+        final Date exp = jwtClaimsSet.getExpirationTime();
+        final Date nbf = jwtClaimsSet.getIssueTime();
+        final Date auth_time = jwtClaimsSet.getDateClaim("auth_time");
+        final String issuer = jwtClaimsSet.getIssuer();
+        final List<String> audiences = jwtClaimsSet.getAudience();
+        LOGGER.debug(
+            "Doing JWT validation, now=["
+                + now
+                + "], typ=["
+                + jwtHeaderType
+                + "], alg=["
+                + jwtHeaderAlgorithm
+                + "], kid=["
+                + jwsHeaderKeyID
+                + "], auth_time=["
+                + auth_time
+                + "], iat=["
+                + iat
+                + "], nbf=["
+                + nbf
+                + "], exp=["
+                + exp
+                + "], issuer=["
+                + issuer
+                + "], audiences=["
+                + audiences
+        );
+
+        // Header "typ" must be "jwt" or null
+        try {
+            JWT_HEADER_TYPE_VERIFIER.verify(jwtHeaderType, null);
+        } catch (Exception e) {
+            throw new Exception("Invalid JWT type [" + jwtHeaderType + "].", e);
+        }
+
+        // Header "alg" must be in the allowed list
+        if ((jwtHeaderAlgorithm == null) || (allowedSignatureAlgorithms.contains(jwtHeaderAlgorithm.getName()) == false)) {
+            throw new Exception(
+                "Rejected signature algorithm ["
+                    + jwtHeaderAlgorithm
+                    + "]. Allowed signature algorithms are ["
+                    + String.join(",", allowedSignatureAlgorithms)
+                    + "]"
+            );
+        }
+
+        // Probable sequence of time claims: auth_time (opt), iat (req), nbf (opt), exp (req)
```

Comment:
Yes. The key word is "probable" is important. Also, this comment helped me visualize the timeline, and decide which time checks are needed versus not needed.

Specifically, I decided to only check each time against wall clock, and not each other. Based on experience with PKI, notBefore and notAfter (aka exp) sometimes include jitter for entropy against predictability. In this case, nbf could sometimes be before iat. Also, it was not clear if skew times should apply when comparing two time claims to each other.

Reply:
I'll rephrase: I realize that this comment added value to you while you were developing the feature, but that's not usually why we add comments in code. I (as a reviewer, maintainer, contributor) see this and I don't have an idea what it means, until I ask you and you elaborate. So my ask is if we think this adds value, lets write it down in a way that is understandable, otherwise lets remove it.

Reply:
Removed comment.